plugin-server/src/main/ingestion-queues/session-recording/snapshot-segmenter.ts

```diff
@@ -0,0 +1,133 @@
+}
+
+export interface RRWebEventSummary {
+    timestamp: number
+    type: number
+    data: RRWebEventSummaryData
+    windowId: string
+}
+
+interface RecordingSegment {
+    kind: 'window' | 'buffer' | 'gap'
+    startTimestamp: number // Epoch time that the segment starts
+    endTimestamp: number // Epoch time that the segment ends
+    durationMs: number
+    windowId?: string
+    isActive: boolean
+}
+
+const isActiveEvent = (event: RRWebEventSummary): boolean => {
+    return event.type === 3 && activeSources.includes(event.data?.source || -1)
+}
+
+const createSegments = (snapshots: RRWebEventSummary[]): RecordingSegment[] => {
+    let segments: RecordingSegment[] = []
+    let activeSegment!: Partial<RecordingSegment>
+    let lastActiveEventTimestamp = 0
+
+    snapshots.forEach((snapshot) => {
+        const eventIsActive = isActiveEvent(snapshot)
+        lastActiveEventTimestamp = eventIsActive ? snapshot.timestamp : lastActiveEventTimestamp
+
+        // When do we create a new segment?
+        // 1. If we don't have one yet
+        let isNewSegment = !activeSegment
+
+        // 2. If it is currently inactive but a new "active" event comes in
+        if (eventIsActive && !activeSegment?.isActive) {
+            isNewSegment = true
+        }
+
+        // 3. If it is currently active but no new active event has been seen for the activity threshold
+        if (activeSegment?.isActive && lastActiveEventTimestamp + ACTIVITY_THRESHOLD_MS < snapshot.timestamp) {
+            isNewSegment = true
+        }
+
+        // 4. If windowId changes we create a new segment
+        if (activeSegment?.windowId !== snapshot.windowId) {
+            isNewSegment = true
+        }
+
+        if (isNewSegment) {
+            if (activeSegment) {
+                segments.push(activeSegment as RecordingSegment)
+            }
+
+            activeSegment = {
+                kind: 'window',
+                startTimestamp: snapshot.timestamp,
+                windowId: snapshot.windowId,
+                isActive: eventIsActive,
+            }
+        }
+
+        activeSegment.endTimestamp = snapshot.timestamp
+    })
+
+    if (activeSegment) {
+        segments.push(activeSegment as RecordingSegment)
+    }
+
+    segments = segments.reduce((acc, segment, index) => {
+        const previousSegment = segments[index - 1]
+        const list = [...acc]
+
+        if (previousSegment && segment.startTimestamp - previousSegment.endTimestamp > 1) {
+            const startTimestamp = previousSegment.endTimestamp + 1
+            const endTimestamp = segment.startTimestamp - 1
+            const gapSegment: Partial<RecordingSegment> = {
+                kind: 'gap',
+                startTimestamp,
+                endTimestamp,
+                isActive: false,
+            }
+
+            list.push(gapSegment as RecordingSegment)
+        }
+
+        list.push(segment)
+
+        return list
+    }, [] as RecordingSegment[])
```

Comment:
In this context we don't care about gaps
```suggestion
    segments = segments.reduce((acc, segment, index) => {
        const previousSegment = segments[index - 1]
        const list = [...acc]

        if (previousSegment && segment.startTimestamp - previousSegment.endTimestamp > 1) {
            const startTimestamp = previousSegment.endTimestamp + 1
            const endTimestamp = segment.startTimestamp - 1
            const gapSegment: Partial<RecordingSegment> = {
                kind: 'gap',
                startTimestamp,
                endTimestamp,
                isActive: false,
            }

            list.push(gapSegment as RecordingSegment)
        }

        list.push(segment)

        return list
    }, [] as RecordingSegment[])
```

Reply:
<img width="275" alt="Screenshot 2023-05-04 at 09 13 32" src="https://user-images.githubusercontent.com/984817/236147592-83285b48-fdd0-45a1-aee4-568ffca2ed66.png">

🙈